tactics/btermdn.ml

```diff
@@ -56,111 +56,86 @@ let decomp sigma t =
   else true
+  &&
+  (match ts with None -> true | Some ts -> TransparentState.is_transparent_constant ts c)
 
-let constr_val_discr env sigma t =
-  let open GlobRef in
-  let c, l = decomp sigma t in
-    match EConstr.kind sigma c with
-    | Ind (ind_sp,u) -> Label(GRLabel (IndRef ind_sp),l)
-    | Construct (cstr_sp,u) -> Label(GRLabel (ConstructRef cstr_sp),l)
-    | Var id -> Label(GRLabel (VarRef id),l)
-    | Const (c, _) ->
-      if evaluable_constant c env then Everything
-      else Label(GRLabel (ConstRef c),l)
-    | _ -> Nothing
+let evaluable_named id env ts = match ts with
+| None -> false
+| Some ts ->
+  Environ.evaluable_named id env && TransparentState.is_transparent_variable ts id
 
-let constr_pat_discr env t =
-  if not (Patternops.occur_meta_pattern t) then
-    None
-  else
-    let open GlobRef in
-    match decomp_pat t with
-    | PRef ((IndRef _) as ref), args
-    | PRef ((ConstructRef _ ) as ref), args -> Some (GRLabel ref,args)
-    | PRef ((VarRef v) as ref), args -> Some(GRLabel ref,args)
-    | PRef ((ConstRef c) as ref), args ->
-      if evaluable_constant c env then None
-      else Some (GRLabel ref, args)
-    | _ -> None
-
-let constr_val_discr_st env sigma ts t =
+let constr_val_discr env sigma ts t =
   let c, l = decomp sigma t in
   let open GlobRef in
     match EConstr.kind sigma c with
     | Const (c,u) ->
-      if evaluable_constant c env && TransparentState.is_transparent_constant ts c then Everything
+      if evaluable_constant c env ts then Everything
       else Label(GRLabel (ConstRef c),l)
     | Ind (ind_sp,u) -> Label(GRLabel (IndRef ind_sp),l)
     | Construct (cstr_sp,u) -> Label(GRLabel (ConstructRef cstr_sp),l)
     | Var id ->
-      if Environ.evaluable_named id env && TransparentState.is_transparent_variable ts id then Everything
+      if evaluable_named id env ts then Everything
       else Label(GRLabel (VarRef id),l)
-    | Prod (n, d, c) -> Label(ProdLabel, [d; c])
+    | Prod (n, d, c) ->
+      if Option.is_empty ts then Nothing else Label(ProdLabel, [d; c])
     | Lambda (n, d, c) ->
-      if List.is_empty l then
-        Label(LambdaLabel, [d; c] @ l)
+      if Option.is_empty ts then Nothing
+      else if List.is_empty l then Label(LambdaLabel, [d; c])
       else Everything
-    | Sort _ -> Label(SortLabel, [])
-    | Evar _ -> Everything
+    | Sort _ -> if Option.is_empty ts then Nothing else Label(SortLabel, [])
+    | Evar _ -> if Option.is_empty ts then Nothing else Everything
     | Rel _ | Meta _ | Cast _ | LetIn _ | App _ | Case _ | Fix _ | CoFix _
     | Proj _ | Int _ | Float _ | Array _ -> Nothing
 
-let constr_pat_discr_st env ts t =
+let constr_pat_discr env ts t =
   let open GlobRef in
-  match decomp_pat t with
+  if Option.is_empty ts && not (Patternops.occur_meta_pattern t) then
+    (* Extremely fishy, and breaks very quickly when fiddled with. Why should
+       **ground** terms match **everything** when non-discriminated??? *)
```

Comment:
I think (at least under some conditions) rewriting is modulo conversion on closed terms?  cc @mattam82

Reply:
Interestingly, the [documentation](https://coq.inria.fr/refman/proofs/automatic-tactics/auto.html#coq:cmd.Create-HintDb) states the exact contrary of the behaviour of this code:
> The legacy implementation (the default one for new databases) uses the DT only on goals without existentials (i.e., auto goals), for non-Immediate hints and does not make use of transparency hints

Literally all of the above is wrong. The DT is only used for terms containing metas, and the first claim contradicts the second one. Furthermore the transparency hints are passed to the unification algorithm even for non-discriminated databases. I believe that the intended meaning of the second part is that the DT retrieval does not use transparency hints, but this is not what the phrasing implies.